pkg/detector/detector.go

```diff
@@ -1007,19 +1006,39 @@ func (d *ResourceDetector) ReconcilePropagationPolicy(key util.QueueKey) error {
 	unstructuredObj, err := d.propagationPolicyLister.Get(ckey.NamespaceKey())
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.Infof("PropagationPolicy(%s) has been removed.", ckey.NamespaceKey())
-			return d.HandlePropagationPolicyDeletion(ckey.Namespace, ckey.Name)
+			return nil
 		}
 		klog.Errorf("Failed to get PropagationPolicy(%s): %v", ckey.NamespaceKey(), err)
 		return err
 	}
 
-	klog.Infof("PropagationPolicy(%s) has been added.", ckey.NamespaceKey())
 	propagationObject := &policyv1alpha1.PropagationPolicy{}
 	if err = helper.ConvertToTypedObject(unstructuredObj, propagationObject); err != nil {
 		klog.Errorf("Failed to convert PropagationPolicy(%s) from unstructured object: %v", ckey.NamespaceKey(), err)
 		return err
 	}
+
+	if !propagationObject.DeletionTimestamp.IsZero() {
+		klog.Infof("PropagationPolicy(%s) has been deleted.", ckey.NamespaceKey())
+		if err = d.HandlePropagationPolicyDeletion(propagationObject.Namespace, propagationObject.Name); err != nil {
+			return err
+		}
+		if controllerutil.RemoveFinalizer(propagationObject, util.PropagationPolicyControllerFinalizer) {
+			if err = d.Client.Update(context.TODO(), propagationObject); err != nil {
+				klog.Errorf("Failed to remove finalizer for PropagationPolicy(%s), err: %v", ckey.NamespaceKey(), err)
+				return err
+			}
+		}
+		return nil
+	}
+
+	klog.Infof("PropagationPolicy(%s) has been added.", ckey.NamespaceKey())
+	if controllerutil.AddFinalizer(propagationObject, util.PropagationPolicyControllerFinalizer) {
```

Comment:
Can the logic of the new finalizer also be added to `karmada-webhook`? I understand that the code here is for compatibility and can be removed in the next version. 

So according to the [comments](https://github.com/karmada-io/karmada/pull/4836#discussion_r1568177480), the policy is always enqueued twice when it is first created, which is not very elegant. When there are too many policies waiting in the queue, the interval between the two times will be large.

Reply:
@RainbowMango @chaunceyjiang

Reply:
> Can the logic of the new finalizer also be added to karmada-webhook?

Yes, we should, please do it at another PR. (split another task?)
But for now, I don't think reconciling twice is a big problem, because the first reconcile is cheap.

Reply:
> Yes, we should, please do it at another PR. (split another task?)

yeah, maybe we don't need add it to karmada-webhook for now, otherwise it would appear redundant. We can completely transplant it to karmada-webhook when we remove this part of the logic in the future. I'll add some comments for it!